### PR TITLE
Add exhaustive tests for compress() and decompress()

### PR DESCRIPTION
**Pull request by DavidBuchanan314**
This tests against vectors I generated [here](https://gist.github.com/DavidBuchanan314/a6fcf8cc465e861916a8eb9f8f233ce5), using Python's Fraction module.

The good news is, the tests pass! 

The bad news is, I don't know golang, so the quality of my code that implements the tests might suck. Let me know what you think heh.

Yours is also the first implementation I've tested these vectors against, aside from my own implementations, so you might want to wait for some additional confirmations.

### CODE REVIEW

**Comment by FiloSottile** [comment]
Sweet, love to get some exhaustive tests.

Do you think you could replicate them with math/big.Rat inline rather than checking in the vectors?

**Comment by DavidBuchanan314** [comment]
Good idea, I converted the tests to use math/big.Rat.

I had some trouble deciding the best way to get a rounded integer result from the Rat, and I ended up with this:

```go
strconv.ParseInt(precise.FloatString(0), 10, 64)
```

It's a bit of a hack, do you know if there's a better way?

File: mlkem768_test.go [hunk 1]
```diff
@@ -79,6 +80,72 @@ func TestDecompressCompress(t *testing.T) {
...
+
+	// FloatString rounds halves away from 0, and our result should always be positive,
+	// so it should work as we expect.
+	// (There's no direct way to convert a Rat to an integer)
+	rounded, err := strconv.ParseInt(precise.FloatString(0), 10, 64)
+	if err != nil {
+		panic(err)
+	}
+
+	return rounded % (1 << d)
```
Comment (FiloSottile) [inline mlkem768_test.go:101]: Why the modulo? I thought the result should always come out to less than 2ᵈ since x < q.

Comment (FiloSottile) [reply mlkem768_test.go:101]: (Similarly below.)

**Review by FiloSottile — APPROVED** [review]
A few nits but I like this! FloatString is a hack, but for tests I think it's fine, and it won't break in the future.

File: mlkem768_test.go (hunk 1 above)
Comment (DavidBuchanan314) [reply mlkem768_test.go:101]: Consider d=1, x=q-1

`(2ᵈ / q) * x` gives `(2/q)*(q-1)`, which evaluates to `1.99...`, rounding up to 2 - which is outside the range of [0..2ᵈ), requiring the reduction

File: mlkem768_test.go [hunk 2]
```diff
@@ -79,6 +80,72 @@ func TestDecompressCompress(t *testing.T) {
 	}
 }
 
+func CompressRat(x, d int64) int64 {
+	if (x < 0) || (x >= q) {
```
[the remark this answers is not in the document]
Comment (DavidBuchanan314) [reply mlkem768_test.go:84]: `gofmt -s` is actually spitting out the same code as-is, but I'll remove them as suggested

File: mlkem768_test.go (hunk 1 above)
Comment (DavidBuchanan314) [reply mlkem768_test.go:101]: I'll add a comment noting this

File: mlkem768_test.go [hunk 3]
```diff
@@ -79,6 +80,72 @@ func TestDecompressCompress(t *testing.T) {
 	}
 }
 
+func CompressRat(x, d int64) int64 {
```
[the remark this answers is not in the document]
Comment (DavidBuchanan314) [reply mlkem768_test.go]: good call, this was just me being lazy to avoid unnecessary casting, but it's definitely better to have the types match

**Comment by DavidBuchanan314** [comment]
Thank you too, for the comprehensive review!

**Comment by FiloSottile** [comment]
@DavidBuchanan314, have you signed the Google Individual CLA (https://cla.developers.google.com/about/google-individual) and do I have your permission to submit this change under the terms of that CLA on your behalf as part of https://go.dev/cl/550215?

**Comment by DavidBuchanan314** [comment]
Signed just now, sounds good to me :)

**Comment by FiloSottile** [comment]
Thank you!

**Pull request merged**